File: gateway_service/routes/transaction.py
"""Маршруты Gateway → Transaction Service (операции по счетам)."""

from uuid import UUID

from fastapi import APIRouter, Depends, Query, Request, status

from shared import schemas
from gateway_service.helpers import forward_request
from gateway_service.middleware import session_token_scheme

router = APIRouter(
	prefix="/accounts",
	tags=["transactions"],
	dependencies=[Depends(session_token_scheme)],
)


@router.post(
	"/{account_id}/deposit",
	response_model=schemas.TransactionMessageResponse,
	status_code=status.HTTP_200_OK,
	summary="Пополнить счёт",
)
async def deposit(
	account_id: UUID,
	payload: schemas.DepositRequest,
	request: Request,
):
	"""Вносит средства на указанный счёт."""
	data = await forward_request(
		request,
		"POST",
		f"/accounts/{account_id}/deposit",
		payload.model_dump(mode="json"),
		service="transaction",
	)
	return data


@router.post(
	"/{account_id}/withdraw",
	response_model=schemas.TransactionMessageResponse,
	status_code=status.HTTP_200_OK,
	summary="Снять со счёта",
)
async def withdraw(
	account_id: UUID,
	payload: schemas.WithdrawalRequest,
	request: Request,
):
	"""Списывает средства со счёта текущего пользователя."""
	data = await forward_request(
		request,
		"POST",
		f"/accounts/{account_id}/withdraw",
		payload.model_dump(mode="json"),
		service="transaction",
	)
	return data


@router.post(
	"/{account_id}/transfer",
	response_model=schemas.TransactionMessageResponse,
	status_code=status.HTTP_200_OK,
	summary="Перевести между счетами",
)
async def transfer(
	account_id: UUID,
	payload: schemas.TransferRequest,
	request: Request,
):
	"""Переводит средства с указанного счёта на другой (свой или чужой)."""
	data = await forward_request(
		request,
		"POST",
		f"/accounts/{account_id}/transfer",
		payload.model_dump(mode="json"),
		service="transaction",
	)
	return data


@router.get(
	"/{account_id}/transactions",
	response_model=schemas.TransactionListResponse,
	status_code=status.HTTP_200_OK,
	summary="История операций",
)
async def transaction_history(
	account_id: UUID,
	request: Request,
	limit: int = Query(20, ge=1, le=100, description="Кол-во записей на странице"),
	offset: int = Query(0, ge=0, description="Смещение"),
	type: str | None = Query(None, description="Тип операции: deposit | withdrawal | transfer"),
	direction: str | None = Query(None, description="Направление: incoming | outgoing"),
):
	"""Возвращает историю операций по счёту с пагинацией и фильтрами."""
	params = f"?limit={limit}&offset={offset}"
	if type:
		params += f"&type={type}"
	if direction:
		params += f"&direction={direction}"

	data = await forward_request(
		request,
		"GET",
		f"/accounts/{account_id}/transactions{params}",
		service="transaction",
	)
	return data

File: gateway_service/routes/transaction.go
package routes

import (
	"fmt"
	"net/http"

	"github.com/labstack/echo/v4"

	"gateway_service/proxy"
)

// TransactionHandler обрабатывает маршруты transaction_service.
type TransactionHandler struct {
	Proxy  *proxy.ServiceClients
	APIKey string
}

// RegisterTransactionRoutes регистрирует маршруты транзакций.
func (h *TransactionHandler) RegisterTransactionRoutes(e *echo.Echo) {
	e.POST("/accounts/:account_id/deposit", h.deposit)
	e.POST("/accounts/:account_id/withdraw", h.withdraw)
	e.POST("/accounts/:account_id/transfer", h.transfer)
	e.GET("/accounts/:account_id/transactions", h.transactionHistory)
}

// deposit godoc
// @Summary     Пополнить счёт
// @Description Вносит средства на указанный счёт.
// @Tags        transactions
// @Security    SessionToken
// @Accept      json
// @Produce     json
// @Param       account_id path string true "UUID счёта"
// @Param       payload body schemas.AmountPayload true "Сумма пополнения"
// @Success     200 {object} map[string]interface{}
// @Failure     401 {object} map[string]string
// @Router      /accounts/{account_id}/deposit [post]
func (h *TransactionHandler) deposit(c echo.Context) error {
	accountID := c.Param("account_id")
	body, _ := readBody(c)
	path := fmt.Sprintf("/accounts/%s/deposit", accountID)
	return h.Proxy.ForwardRaw(c, http.MethodPost, path, body, "transaction", h.APIKey)
}

// withdraw godoc
// @Summary     Снять со счёта
// @Description Списывает средства со счёта текущего пользователя.
// @Tags        transactions
// @Security    SessionToken
// @Accept      json
// @Produce     json
// @Param       account_id path string true "UUID счёта"
// @Param       payload body schemas.AmountPayload true "Сумма снятия"
// @Success     200 {object} map[string]interface{}
// @Failure     401 {object} map[string]string
// @Router      /accounts/{account_id}/withdraw [post]
func (h *TransactionHandler) withdraw(c echo.Context) error {
	accountID := c.Param("account_id")
	body, _ := readBody(c)
	path := fmt.Sprintf("/accounts/%s/withdraw", accountID)
	return h.Proxy.ForwardRaw(c, http.MethodPost, path, body, "transaction", h.APIKey)
}

// transfer godoc
// @Summary     Перевести между счетами
// @Description Переводит средства с указанного счёта на другой (свой или чужой).
// @Tags        transactions
// @Security    SessionToken
// @Accept      json
// @Produce     json
// @Param       account_id path string true "UUID счёта-отправителя"
// @Param       payload body schemas.TransferRequest true "Данные перевода"
// @Success     200 {object} map[string]interface{}
// @Failure     401 {object} map[string]string
// @Router      /accounts/{account_id}/transfer [post]
func (h *TransactionHandler) transfer(c echo.Context) error {
	accountID := c.Param("account_id")
	body, _ := readBody(c)
	path := fmt.Sprintf("/accounts/%s/transfer", accountID)
	return h.Proxy.ForwardRaw(c, http.MethodPost, path, body, "transaction", h.APIKey)
}

// transactionHistory godoc
// @Summary     История операций
// @Description Возвращает историю операций по счёту с пагинацией и фильтрами.
// @Tags        transactions
// @Security    SessionToken
// @Produce     json
// @Param       account_id path string true "UUID счёта"
// @Param       limit query int false "Кол-во записей (по умолчанию 20)" default(20)
// @Param       offset query int false "Смещение" default(0)
// @Param       type query string false "Тип операции: deposit | withdrawal | transfer"
// @Param       direction query string false "Направление: incoming | outgoing"
// @Success     200 {object} map[string]interface{}
// @Failure     401 {object} map[string]string
// @Router      /accounts/{account_id}/transactions [get]
func (h *TransactionHandler) transactionHistory(c echo.Context) error {
	accountID := c.Param("account_id")

	queryString := c.QueryString()
	path := fmt.Sprintf("/accounts/%s/transactions", accountID)
	if queryString != "" {
		path += "?" + queryString
	}
	return h.Proxy.ForwardRaw(c, http.MethodGet, path, nil, "transaction", h.APIKey)
}
